quinn/src/send_stream.rs

```diff
@@ -199,27 +202,16 @@ impl SendStream {
     /// For a variety of reasons, the peer may not send acknowledgements immediately upon receiving
     /// data. As such, relying on `stopped` to know when the peer has read a stream to completion
     /// may introduce more latency than using an application-level response of some sort.
-    pub async fn stopped(&mut self) -> Result<Option<VarInt>, StoppedError> {
-        Stopped { stream: self }.await
-    }
-
-    fn poll_stopped(&mut self, cx: &mut Context) -> Poll<Result<Option<VarInt>, StoppedError>> {
-        let mut conn = self.conn.state.lock("SendStream::poll_stopped");
-
-        if self.is_0rtt {
-            conn.check_0rtt()
-                .map_err(|()| StoppedError::ZeroRttRejected)?;
-        }
-
-        match conn.inner.send_stream(self.stream).stopped() {
-            Err(_) => Poll::Ready(Ok(None)),
-            Ok(Some(error_code)) => Poll::Ready(Ok(Some(error_code))),
-            Ok(None) => {
-                if let Some(e) = &conn.error {
-                    return Poll::Ready(Err(e.clone().into()));
+    pub fn stopped(
+        &self,
+    ) -> impl Future<Output = Result<Option<VarInt>, StoppedError>> + Send + Sync + 'static {
+        let (conn, stream, is_0rtt) = (self.conn.clone(), self.stream, self.is_0rtt);
+        async move {
+            loop {
+                match stopped_or_notify(&conn, stream, is_0rtt) {
+                    ControlFlow::Break(res) => return res,
+                    ControlFlow::Continue(notify) => notify.notified().await,
```

Comment:
On review, this is racy. `notify_waiters` only wakes currently registered tasks, but might be called in between the lock being released and `notified` being called. The `Notified` future *must* be constructed (but not polled!) before the connection lock is released.

Reply:
Good catch. I refactored it so that the `Notified` future is created while the lock is still being held. Due to `Notified` being lifetime-bound to `Notify`, this needed a manual `super let` construction, see the comment in the code.